### ISSUE TITLE
`Get-Variable -Scope <number>` should (possibly) return the (inherited) read scope

### ISSUE THREAD

**Issue by iRon7** [PowerShell/PowerShell#18177]
### Summary of the new feature / enhancement

As a user I would expect the `Get-Variable -Scope <number>` to return  the (inherited) read scope of the concerned variable.
(or at least have an easy way to do so with e.g. an extra parameter.)

```PowerShell
function Child {
    Write-Host 'Parent DebugPreference:' (Get-Variable 'DebugPreference' -Scope 1 -ValueOnly)
}

function Parent {
    Write-Host 'Parent DebugPreference:' $DebugPreference
    Child
}
Parent
```

### Expected:
```Console
Parent DebugPreference: SilentlyContinue
Parent DebugPreference: SilentlyContinue
```

### Actual:
```Console
Parent DebugPreference: SilentlyContinue
Get-Variable: C:\Users\Gebruiker\Scripts\Test-Scope\Test-Scope.ps1:10
Line |
  10 |  … ebugPreference:' (Get-Variable 'DebugPreference' -Scope 1 -ValueOnly)
     |                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot find a variable with the name 'DebugPreference'.
```

### Proposed technical implementation details (workaround)
```PowerShell
function Child {
    for ($i = 1; $i -lt (Get-PSCallStack).Count; $i++) {
        $ParentDebugPreference = Get-Variable 'DebugPreference' -Scope $i -ValueOnly -ErrorAction SilentlyContinue
        if ($Null -ne $ParentDebugPreference) { break }
    }
    Write-Host 'Parent DebugPreference:' $ParentDebugPreference
}

function Parent {
    Write-Host 'Parent DebugPreference:' $DebugPreference
    Child
}
Parent
```

**Comment by 237dmitry** [comment]
In have to set variable in `Parent` function:

```powershell
function Child {
    Get-Variable DebugPreference -Scope 1 -ValueOnly
}

function Parent {
    $DebugPreference
    $DebugPreference = 'Stop' 
    Child
}

Parent

# SilentlyContinue
# Stop
```

**Comment by iRon7** [comment]
@237dmitry
> have to set variable in `Parent` function

Yes, I know but:
1. As a user, I don't expected that to be required (I don't need to do that for the `Write-Host 'Parent DebugPreference:' $DebugPreference` in the parent function either)
2. [`Get-...`](https://learn.microsoft.com/powershell/scripting/developer/cmdlet/approved-verbs-for-windows-powershell-commands?view=powershell-7.2#get-vs-read) suggest that a user will **read** that variable in the specific scope (which might simply be inherited from another parent)
3. Users might not have control over the parent function

Note that `$DebugPreference` is just an easy example, but the request/issue counts for any variable set in a higher scope.

**Comment by 237dmitry** [comment]
If you didn't set the variable then the parent scope will remain empty (there are no variables in it). Therefore, `Get-Variable` cannot return a value, does not find a variable in the parent scope.

**Comment by iRon7** [comment]
> If you didn't set the variable then the parent scope will remain empty (there are no variables in it). Therefore, `Get-Variable` cannot return a value, does not find a variable in the parent scope.

I understand that, therefore I am requesting an enhancement to easily *read* the (inherited read) variable in the parent scope just as I would be able to retrieve that variable value in the actual parent scope by its `$<name>` (as e.g. `$DebugPreference`) without changing the parent function, similar to the **Proposed technical implementation details (workaround)**

**Comment by 237dmitry** [comment]
It's easier to call `Child` function with parameter.
```powershell
$Test = 'not in Parent'            
function Child ($ParentVariables) {
    $DebugPreference      
    $ErrorActionPreference
    $Test
}                
                                        
function Parent {                         
    $DebugPreference = 'Continue'  
    $ErrorActionPreference = 'Stop'          
    $Test = 'test'                           
    $ScopeVariables = Get-Variable * -Scope 0
    Child $ScopeVariables
}     
      
Parent
$Test
```

**Comment by iRon7** [comment]
> It's easier to call Child function with parameter.

As said, I am only in control of the child function (and not in control of the parent function). In my specific case, it concerns a "`Write-Log`" function where I want to get details of specific available (*inherited* preference) variables of the caller but I imaging there are other similar use cases for this request.

**Comment by 237dmitry** [comment]
> As said, I am only in control of the child function (and not in control of the parent function)

Get WHOLE set of parent scope and use variables you need:
```powershell
$Test = 'not in Parent'
function Child {
    $null = Get-Variable * -Scope 1
    $Test
}

function Parent {
    $Test = 'in Parent Scope'               
    Child                          
}                 
         
Parent
$Test
```

**Comment by iRon7** [comment]
> Get WHOLE set of parent scope and use variables you need:

That doesn't resolve anything for what I am requesting for as there *might or **might not*** be a `$Test` assignment (change) in the `Parent` function.

> [Unless you explicitly make the items private, the items in the parent scope are available to the child scope.](https://learn.microsoft.com/powershell/module/microsoft.powershell.core/about/about_scopes?view=powershell-7.2#parent-and-child-scopes)

**Wishful thinking**
e.g. using a additional or different parameter (like `-Available`)

```PowerShell
$Test = 'Test'
function Child {
    $Test = 'I do not want the $Test in this scope'
    Write-Host (Get-Variable Test -Available 1 -ValueOnly) # I want the current $Test ('Test') _available_ in the parent scope
}

function Parent {
    Write-Host '$Test is available in this scope:' $Test          
    Child                          
}                 
         
Parent
```

**Comment by SeeminglyScience** [comment]
The way I'd recommend implementing this is to place your function in a different module. e.g.

```powershell
New-Module {
    function Log {
        [CmdletBinding()]
        param()
        end {
            # This won't be found
            $test = 'no'

            # `$PSCmdlet.SessionState` refers to the *caller's* session state
            $PSCmdlet.SessionState.PSVariable.Get('test')
        }
    }
} | Import-Module

$test = 'yes'
function Caller {
    Log
}

Caller
```

**Comment by 237dmitry** [comment]
> `Get-Variable Test -Available 1 -ValueOnly`

What this switch has to return, variable's value or boolean? What the difference with:
```powershell
Get-Variable Test -Scope 1 -ValueOnly -ErrorAction 'SilentlyContinue'
```
```powershell
function Child {
   $null = Get-Variable Test -Scope 1 -ValueOnly -ErrorAction 0 || Set-Variable Test 'child'
   $Test
}       
 
function Parent {
   $Test = 'parent'  
   Child            
}                 
Parent

```

**Comment by iRon7** [comment]
@SeeminglyScience,

> The way I'd recommend implementing this is to place your function in a different module

Thanks for the suggestion, this is indeed the way it should be implemented for my use case.

Unfortunately, the environment concerned is a little more complex as the log script is a part of a larger module with about 20 other scripts that are imported. The CI/CD pipeline that is setup (pester) tests each function by dot-sourcing it where your suggestion acts slightly different than my workaround.

Anyways, I am still interested in your view of this issue, whether:
* The default **`Get`**`-Variable` behavior could be considered incorrect
  * if yes, I presume the default behavior won't change as is might be a considered as a break change even it currently generates an error in the case the variable isn't (re)set in the concerned scope.
 * If it could be fixed/improved with an extra parameter

**Comment by jhoneill** [comment]
> Anyways, I am still interested in your view of this issue, whether:
> 
> * The default **`Get`**`-Variable` behavior could be considered incorrect
>   
>   * if yes, I presume the default behavior won't change as is might be a considered as a break change even it currently generates an error in the case the variable isn't (re)set in the concerned scope.
> * If `Get-Variable` could be fixed/improved with an extra parameter


I am trying to figure out _what the actual requirement is_  and if you need to use `get-variable` at all. If you want the inherited $x won't `$x` provide what you want in almost all cases ? 
Some example code to show why I think `Get-Variable` isn't needed. 
```
function a {
$x = 1 
"Local x in a is '$local:x' ; x is '$x'"
b
}

function b {
"Local x in b is '$local:x' ; x is '$x'"
c
}

function c {
"Local x in c is '$local:x' ; x is '$x'"
}

a
Local x in a is '1' ; x is '1'
Local x in b is '' ; x is '1'
Local x in c is '' ; x is '1'
```

This shows that in B and C, we can test for  "X was (not) set locally", "X has a value" therefore it's simple to get to "X was inherited", _without caring where from_  If X is a parameter or a preference variable set via a parameter, I'm trying to think of a real-world use case for wanting to know what it was before it was set here. "$x from all higher scopes but not this one" is something I have never needed myself, so I'm having trouble imagining a use for it.   

In a more complicated example.
```
function c {
## other code would be here
    if     ($local:x )    {"x was set in This function"}    
    elseif ($x -and     (Get-Variable -Name X -Scope 1 -ErrorAction SilentlyContinue) ) {
          "x was set in this function's parent"
    }
    elseif ($global:x ) {
          "x was set in the Global scope"
    }
    else  { "x was set higher than this functions parent but not globally"}
}

PS> a
Local x in a is '1' ; x is '1'
Local x in b is '' ; x is '1'
x was set higher than this functions parent but not globally
```

This shows an example where we don't want the present behaviour of `Get-Variable` to change (we want to know if a variable was or was not set in scope 1)  but **maybe** a "AND ABOVE"  switch could be helpful - but as a I said my imagination isn't giving me any cases. 


As @SeeminglyScience  says variables don't inherit out of modules and I have a post about that, and how preference variables can be passed here:    
https://devblogs.microsoft.com/powershell-community/how-to-have-more-control-of-preferences-in-functions-and-the-role-of-modules-on-inheritance/   it's not clear from the question if you are trying to solve a problem with `$DebugPreference` but that post might help if you are.

**Comment by jhoneill** [comment]
The number of times I've needed `Get-Variable` at all has been very small, and I've used `Set-variable` with `-Scope` to do something similar to `-OutVariable` / `-ErrorVariable` to leave something in a variable selected by the caller where that is different from the _output_.   

Because I have solved this problem for one case with the stuff in that post I would probably use hidden parameters prefixed with "parent" or similar to get the value when the function was called. That doesn't mean it is the perfect way to do it, but knowing a a solution for almost the same problem I would adapt it. You have another work round, and this is very specific - I wouldn't expect the extra parameter to be rejected as a bad idea, but I don't see it being treated as important enough to get time allocated to it.
